operator/controllers/resources/user_controller_test.go

```diff
@@ -65,35 +69,163 @@ func TestUserCreation(t *testing.T) {
 	})
 
 	// The user is deleted in K8S
-	k8sDeleteUser(ctx, t, k8sClient, userName, ns.Name)
+	k8sDeleteUser(ctx, t, setup.k8sClient, userName, setup.namespace.Name)
 
 	fastEventually(t, func() bool {
-		_, err := tClient.GetUser(userName, false)
+		_, err := setup.tClient.GetUser(userName, false)
 		return trace.IsNotFound(err)
 	})
 }
-
-func TestUserDeletionDrift(t *testing.T) {
+func TestUserCreationFromYAML(t *testing.T) {
 	ctx := context.Background()
+	setup := setupKubernetesAndTeleport(t, ctx)
+	_ = teleportCreateDummyRole(ctx, t, "a", setup.tClient)
+	tests := []struct {
+		name         string
+		userSpecYAML string
+		shouldFail   bool
+		expectedSpec *types.UserSpecV2
+	}{
+		{
+			name: "Valid user without traits",
+			userSpecYAML: `
+roles:
+  - a
+`,
+			shouldFail: false,
+			expectedSpec: &types.UserSpecV2{
+				Roles: []string{"a"},
+			},
+		},
+		{
+			name: "Valid user with trait (list with single element)",
+			userSpecYAML: `
+roles:
+  - a
+traits:
+  'foo': ['bar']
+`,
+			shouldFail: false,
+			expectedSpec: &types.UserSpecV2{
+				Roles: []string{"a"},
+				Traits: map[string][]string{
+					"foo": []string{"bar"},
+				},
+			},
+		},
+		{
+			name: "Valid user with traits (list with multiple element)",
+			userSpecYAML: `
+roles:
+  - a
+traits:
+  'foo': ['bar', 'baz']
+`,
+			shouldFail: false,
+			expectedSpec: &types.UserSpecV2{
+				Roles: []string{"a"},
+				Traits: map[string][]string{
+					"foo": []string{"bar", "baz"},
+				},
+			},
+		},
+		{
+			name: "Invalid user with non-existing role",
+			userSpecYAML: `
+roles:
+  - does-not-exist
+traits:
+  'foo': ['bar', 'baz']
+`,
+			shouldFail:   true,
+			expectedSpec: nil,
+		},
+	}
 
-	teleportServer, operatorName := defaultTeleportServiceConfig(t)
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			// Creating the Kubernetes resource. We are using an untyped client to be able to create invalid resources.
+			userManifest := map[string]interface{}{}
+			err := yaml.Unmarshal([]byte(tc.userSpecYAML), &userManifest)
+			require.NoError(t, err)
+
+			userName := validRandomResourceName("user-")
+
+			obj := getUnstructuredObjectFromGVK(TeleportUserGVK)
+			obj.Object["spec"] = userManifest
+			obj.SetName(userName)
+			obj.SetNamespace(setup.namespace.Name)
+			err = setup.k8sClient.Create(ctx, obj)
+			require.NoError(t, err)
+
+			// If failure is expected we should not see the resource in Teleport
+			if tc.shouldFail {
+				// We wait 1 second to ensure reconciliation happened
+				time.Sleep(time.Second)
```

Comment:
As the other thread said, let's try and refactor the Sleeps out. This is flakiness waiting to happen.